jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterConcatIterator.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.sparql.engine.iterator;
+
+import org.apache.jena.atlas.io.IndentedWriter;
+import org.apache.jena.atlas.lib.Lib;
+import org.apache.jena.sparql.engine.ExecutionContext;
+import org.apache.jena.sparql.engine.QueryIterator;
+import org.apache.jena.sparql.engine.binding.Binding;
+import org.apache.jena.sparql.serializer.SerializationContext;
+
+import java.util.*;
+
+
+/**
+ * A query iterator that joins two or more iterators into a single iterator. */ 
+
+public class QueryIterConcatIterator extends QueryIter
```

Comment:
@afs Is there not already a query iterator that supported this pattern?

I see a `QueryIterConcat` referenced in the code deletions so why was that not suitable?  Could that have been modified to solve the issue rather than introducing a new class

Reply:
the QueryIterConcat can be used to make a materialised list of many QueryIterators. In this case, the (nested) QueryIterators themselves should be constructed lazily, that's why I didn't see a good way to make use of the existing class

Reply:
Isn't that then a flatmap? 
`QueryIterRepeatApply` provides flatmap for `QueryIterator`.

Reply:
should I delete this class and replace it with QueryIterPlainWrapper.create(Iter.flatMap(...))?

Reply:
`QueryIterRepeatApply` if you want cancellation at that granularity.

Reply:
sorry, I don't understand. RepeatApply is an abstract class that takes a QueryIterator and repeatedly calls nextStage, how will that help here?

Reply:
You can rearrange the code in `PathLib` along the lines of first turning the Nodes to Bindings, and then wrap that as a QueryIterator which serves as input to QueryIterRepeatApply:

```java
Iterator<Node> itNode = determineUngroundedStartingSet(graph, path, execCxt) ;
Iterator<Binding> itBinding = Iter.iter(itNode).map(n -> BindingFactory.binding(binding, sVar, n));
QueryIterator lhs = QueryIterPlainWrapper.wrap(itBinding);
QueryIterator combined = new QueryIterRepeatApply(lhs) {
  QueryIterator nextStage() {
                Iterator<Node> pathIter = PathEval.eval(graph, b2.get(sVar), path, execCxt.getContext());
            return evalGroundedOneEnd(b2, pathIter, oVar, execCxt);
  }
};
```

Reply:
would that have any advantage over or be significantly differnt from QueryIterPlainWrapper(Iter.flatMap)?

Reply:
> You can rearrange the code in `PathLib`

There is also `QueryIterExtendByVar`.

Reply:
hummmm... I can see how the execUngroundedPath could be rewritten with ExtendByVar + implementing RepeatApply but for the  execUngroundedPathSameVar the code is more cumbersome as we want to go from Node -> (x, node) -> (x, binding) -> QueryIterator but the RepeatApply only takes a binding

Reply:
> the difference between cancel() and close() 

`cancel` is async e.g. query timeout. `close` is called from creating thread a a controlled point.

Reply:
https://gist.github.com/afs/cb8e4d806d4efd9239682ec9220f936f